### ISSUE TITLE
istio-ingressgateway randomly doesn't return certificate for request after 1.3.x upgrade

### ISSUE THREAD

**Issue by denniswebb** [istio/istio#17660]
**Bug description**
After upgrading to Istio 1.3.1 (also occurs on 1.3.0) we will get random SSL connection errors to resources behind the ingressgateway. Through Cloudflare, these are returned to the user as 525 errors, but going straight to the ELB with curl the error returned by curl is **curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to httpbin.test:443**

Our clusters are hosted on EKS, and the issue can easily be recreated.

Please see the following asciinema for a complete recreation of the issue. Curl request 170 is where the error occurs.
https://asciinema.org/a/qMBH6IBOjma6vqLXNNgoqYdyr

The following gist is used to deploy httpbin and other resources required: https://gist.githubusercontent.com/denniswebb/23957c9bf11fa9fa856fe2b8f0557ab1/raw/098eb9e7cf2999445e94470ea8c59578a6b3e7a3/istio-libressl.yaml

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
All requests to the ingressgateway should return a valid certificate information.

**Steps to reproduce the bug**
1. Create a new eks cluster
1. Install istio 1.3.1 using the value `gateways.istio-ingressgateway.sds.enabled=true`
1. Expose any service using a certificate through the ingressgateway
1. Continuous curl requests until the LibreSSL error message is returned.
 
**Version (include the output of `istioctl version --remote` and `kubectl version`)**
```
istioctl version --remote
client version: unknown
control plane version: 1.3.1

kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-19T13:57:45Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.10-eks-5ac0f1", GitCommit:"5ac0f1d9ab2c254ea2b0ce3534fd72932094c6e1", GitTreeState:"clean", BuildDate:"2019-08-20T22:39:46Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
`helm install istio.io/istio --name istio --namespace istio-system --set-string gateways.istio-ingressgateway.sds.enabled=true`

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS EKS

Please see the above video and gist to view/reproduce issue. I can 100% confirm we never experienced this issue on 1.2.2 releases.

I can not reproduce locally using docker-desktop on Mac, but I'm not 100% convinced given enough requests and time it could happen.

**Comment by nesl247** [comment]
We just ran into this issue as well.

**Comment by denniswebb** [comment]
@neal247 You on EKS or AWS?

**Comment by nesl247** [comment]
GKE using the istio helm chart.

**Comment by nesl247** [comment]
Same here.

**Comment by denniswebb** [comment]
As a test, I just modified the deployment for **istio-ingressgateway** and changed the image tag for **istio/proxyv2** to **1.2.7** and it is working perfect as expected with over 10,000 requests made.

**Comment by chadlwilson** [comment]
Just a suggestion - might be worth highlighting that you are using cert-manager w/ SDS here. Until I watched your Asciinema vid and looked into the gist, I personally wasn't clear that this is the configuration you were running. Might help get the attention of the right folks?

**Comment by nesl247** [comment]
I'll confirm that we are using cert-manager w/ SDS ourselves as well.

**Comment by cdmurph32** [comment]
I'm seeing this without SDS.  Roughly 1 in 1000 POST and PUT calls fail.
`javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake`
We don't have this issue with Istio 1.1.

**Comment by denniswebb** [comment]
@cdmurph32 Can you help confirm that if you use a proxy image for the istio-ingressgateway deployment of `1.2.7` that the issue goes away also. We've had no issues since rolling back just that one component to an older release.

**Comment by cdmurph32** [comment]
@denniswebb 
> @cdmurph32 Can you help confirm that if you use a proxy image for the istio-ingressgateway deployment of `1.2.7` that the issue goes away also. We've had no issues since rolling back just that one component to an older release.

I can confirm that I do not have the issue with version 1.2.7.  Unfortunately for me, I really need the Envoy option `preserve_external_request_id`, which is only available in 1.3.

**Comment by lambdai** [comment]
@denniswebb has the config that https is involved. I suspect tls_inspector is wrongly determine tls traffic as non-tls.

Is there a config_dump so that I can confirm
1. tls_inspector is in the config
2. there is a limited listener timeout
Also,
What is the LDS config update frequency on istio-ingressgateway? You can `kubectl -n istio-system exec -it deploy/istio-ingressgateway -- curl localhost:15000/stats |grep -i lds` and see if it's more than you expected

**Comment by PiotrSikora** [comment]
I'm pretty sure this is the listener filters timeout set too low, can we set it to something more reasonable? Is there an easy way to configure that, @rshriram?

**Comment by lambdai** [comment]
The default listener timeout is 0.1s, which is a small timeout crossing cluster/region. The traffic may failed to be determined as TLS. The follow up depends on the ingress listener config. I will try reproduce once I fix another bug on this and I will get back later

**Comment by yxue** [comment]
We can specify a larger value for this timeout https://github.com/istio/istio/blob/0993a09a47ecce2de3ad31f6cf001066c7ef29af/install/kubernetes/helm/istio/values.yaml#L218
when install the pilot.

**Comment by lambdai** [comment]
I think it a good start to increase the value to 1s or 500s and see if it is mitigated.

**Comment by denniswebb** [comment]
I ran some more tests using the latest images and had some success with a 1.4 build. This happened after some guidance on Istio slack

`gcr.io/istio-release/proxyv2:release-1.3-20191029-10-16`: It took many attempts, but after almost 1,800 requests I got a failure
```
1787
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to httpbin.test:443
1788
```
I re-ran the test again with timestamps because I thought maybe it was a 10-minute thing since 1,800 was about 10 minutes. But on a fresh redeploy of the same image and a new istio-ingressgateway pod, it occured in just 637 requests and about 210 seconds.
```
637
Tue Oct 29 21:06:39 CDT 2019
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to httpbin.test:443
638
Tue Oct 29 21:06:40 CDT 2019
```

`gcr.io/istio-testing/proxyv2:1.4-alpha.a3b0c0f3af66ad3255d2efa1551bbb68aaed9fc2`: **Good news!** No issues after 4,000k requests over 22 minutes.

**Comment by duderino** [comment]
@denniswebb could you try a `protocolDetectionTimeout: 500s` on 1.3?  If the request failures go away that both isolate the issue and give an immediate workaround.

**Comment by rshriram** [comment]
we should just disable the timeout for gateways completely

**Comment by yxue** [comment]
I have a PR for fixing this. https://github.com/istio/istio/pull/18463

**Comment by rshriram** [comment]
What would be ideal is to never even have the TLS inspector in Envoy for things like gateways. It makes no sense to enable TLS sniffing filters for a gateway. Disabling the inspector causes a ton of spam in the envoy logs though https://github.com/istio/istio/issues/7601 which should ideally be fixed rather than us putting a bandaid over a serious issue like this. We had enabled the TLS inspector just to shut envoy up from spamming [https://github.com/istio/istio/pull/8419]

**Comment by rshriram** [comment]
I say user consent because folks from Yahoo! have requested TLS sniffing at the gateway [ #18179 ] but that is explicit intent. So, @yxue 's PR #18463 will just be a temporary bandaid until it gets reverted completely to support TLS sniffing - and then this problem will return.

**Comment by PiotrSikora** [comment]
@rshriram TLS inspector is installed only when needed by the filter chain rules, and it's going to be injected when needed, even when not explicitly configured.

We either misconfigure the filter chain rules (which trigger TLS inspector warning when it's not necessary), or we need TLS inspector, in case your point is moot.

The real issue is introduction of the timeout (with an extremely low default) for cases when we have no business of bypassing the TLS inspector.

**Comment by denniswebb** [comment]
I just tested, and installing with `--set-string global.proxy.protocolDetectionTimeout=500s` fixed the issue.

I then tested with that Timeout set to just 5ms, and got the LibreSSL error for almost 50% of the calls.

Thanks @duderino for the answer.

Should I leave this open until a fix has been implemented for default installs?

**Comment by PiotrSikora** [comment]
@denniswebb yes, please leave it open. And thanks for verifying the fix!

**Comment by incfly** [comment]
Just curious, what's our original use case to put the `tls_inspector`? I saw Yahoo folks' FR is still open, which means some other existing use case requiring us to add tls_inspector there.

**Comment by howardjohn** [comment]
1.3 backport https://github.com/istio/istio/pull/18526 will be in 1.3.5

**Comment by howardjohn** [comment]
This is fixed, will be released shortly in 1.3.5 and in 1.4.0
